File: exp-07/q1-fractional-knapsack/solution.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <vector>
using namespace std;

struct Item {
    int weight;
    int profit;
};

int main() {
    int n;
    double W;
    cin >> n >> W;

    vector<Item> items(n);
    for (int i = 0; i < n; i++) {
        cin >> items[i].weight >> items[i].profit;
    }

    sort(items.begin(), items.end(), [](const Item &a, const Item &b) {
        return (double)a.profit / a.weight > (double)b.profit / b.weight;
    });

    double total = 0.0;
    for (const Item &item : items) {
        if (item.weight <= W) {
            total += item.profit;
            W -= item.weight;
        } else {
            total += item.profit * (W / item.weight);
            W = 0;
        }
    }

    cout << fixed << setprecision(2) << total << endl;
    return 0;
}

File: exp-07/q1-fractional-knapsack/Solution.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

public class Solution {
    static class Item {
        int weight;
        int profit;

        Item(int weight, int profit) {
            this.weight = weight;
            this.profit = profit;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        double W = sc.nextDouble();

        Item[] items = new Item[n];
        for (int i = 0; i < n; i++) {
            items[i] = new Item(sc.nextInt(), sc.nextInt());
        }

        Arrays.sort(items, new Comparator<Item>() {
            public int compare(Item a, Item b) {
                double r1 = (double) a.profit / a.weight;
                double r2 = (double) b.profit / b.weight;
                return Double.compare(r2, r1);
            }
        });

        double total = 0.0;
        for (int i = 0; i < n && W > 0; i++) {
            if (items[i].weight <= W) {
                total += items[i].profit;
                W -= items[i].weight;
            } else {
                total += items[i].profit * (W / items[i].weight);
                W = 0;
            }
        }

        System.out.printf("%.2f%n", total);
        sc.close();
    }
}